core/src/saros/concurrent/management/ConcurrentDocumentClient.java

```diff
@@ -123,16 +160,55 @@ private IActivity receiveChecksum(ChecksumActivity activity) {
     return activity;
   }
 
-  /** Used to remove JupiterClientDocuments for deleted files */
-  private final IActivityReceiver clientReceiver =
-      new IActivityReceiver() {
-        @Override
-        public void receive(FileActivity fileActivity) {
-          if (fileActivity.getType() == FileActivity.Type.REMOVED) {
-            jupiterClient.reset(fileActivity.getPath());
-          }
-        }
-      };
+  /**
+   * Removes the created resource from the held set of deleted shared resources. This causes
+   * subsequent received activities for the resources to be executed again.
+   *
+   * <p>Does nothing if the passed file activity does not have the type {@link Type#CREATED} or
+   * {@link Type#MOVED}.
+   *
+   * @param fileActivity the file activity to handle
+   * @see #deletedResources
+   * @see #handleResourceDeletion(FileActivity)
+   */
+  private void handleResourceCreation(FileActivity fileActivity) {
+    if (fileActivity.getType() == Type.MOVED) {
+      SPath addedFile = fileActivity.getPath();
+
+      deletedResources.remove(addedFile);
+    } else if (fileActivity.getType() == Type.CREATED) {
+      SPath addedFile = fileActivity.getPath();
+
+      deletedResources.remove(addedFile);
+    }
+  }
+
+  /**
+   * Resets the jupiter client for the deleted resource and adds it to the held set of deleted
+   * shared resources. This causes subsequent received activities for the resource to be dropped
+   * until it is created again.
+   *
+   * <p>Does nothing if the passed file activity does not have the type {@link Type#REMOVED} or
+   * {@link Type#MOVED}.
+   *
+   * @param fileActivity the file activity to handle
+   * @see #deletedResources
+   * @see #handleResourceCreation(FileActivity)
+   */
+  private void handleResourceDeletion(FileActivity fileActivity) {
+    if (fileActivity.getType() == Type.REMOVED) {
+      SPath removedFile = fileActivity.getPath();
+
+      deletedResources.add(removedFile);
+      jupiterClient.reset(removedFile);
+
+    } else if (fileActivity.getType() == Type.MOVED) {
+      SPath removedFile = fileActivity.getOldPath();
+
+      deletedResources.add(removedFile);
+      jupiterClient.reset(removedFile);
```

Comment:
And now you send the activity and in the meantime you receive and JupiterActivity for such file, what will happen ?

Reply:
Such activities should be filtered out by the logic in [l.124](https://github.com/saros-project/saros/pull/714/files#diff-cfaccc9315d374b587ed16f4876e129fR124).